src/main/scala/uk/gov/nationalarchives/consignmentexport/Main.scala

```diff
@@ -21,39 +23,59 @@ object Main extends CommandIOApp("tdr-consignment-export", "Exports tdr files in
+          bashCommands = BashCommands()
+          graphQlApi = GraphQlApi(config.api.url, config.auth.url)
+          keycloakClient = KeycloakClient(config)
+          s3Files = S3Files(S3Utils(s3Async))
+          bagit = Bagit()
+          validator = Validator(consignmentId)
+          //Export datetime generated as value needed in bag metadata and DB table
+          //Cannot use the value from DB table in bag metadata, as bag metadata created before bagging
+          //and cannot update DB until bag creation successfully completed
+          exportDatetime = ZonedDateTime.now(ZoneOffset.UTC)
+          consignmentResult <- graphQlApi.getConsignmentMetadata(config, consignmentId)
+          consignmentData <- IO.fromEither(validator.validateConsignmentResult(consignmentResult))
+          _ <- IO.fromEither(validator.validateConsignmentHasFiles(consignmentData))
+          bagMetadata <- BagMetadata(keycloakClient).generateMetadata(consignmentId, consignmentData, exportDatetime)
+          validatedFileMetadata <- IO.fromEither(validator.extractFileMetadata(consignmentData.files))
+          _ <- s3Files.downloadFiles(validatedFileMetadata, config.s3.cleanBucket, consignmentId, basePath)
+          bag <- bagit.createBag(consignmentId, basePath, bagMetadata)
+          fileMetadataCsv <- BagAdditionalFiles(bag.getRootDir).createFileMetadataCsv(validatedFileMetadata)
+          checksums <- ChecksumCalculator().calculateChecksums(fileMetadataCsv)
+          _ <- bagit.writeTagManifestRows(bag, checksums)
+          // The owner and group in the below command have no effect on the file permissions. It just makes tar idempotent
+          _ <- bashCommands.runCommand(s"tar --sort=name --owner=root:0 --group=root:0 --mtime ${java.time.LocalDate.now.toString} -C $basePath -c ./$consignmentId | gzip -n > $tarPath")
+          _ <- bashCommands.runCommand(s"sha256sum $tarPath > $tarPath.sha256")
+          _ <- s3Files.uploadFiles(config.s3.outputBucket, consignmentId, tarPath)
+          _ <- graphQlApi.updateExportLocation(config, consignmentId, s"s3://${config.s3.outputBucket}/$consignmentId.tar.gz", exportDatetime)
+          _ <- stepFunction.publishSuccess(taskToken, ExportOutput(consignmentData.userid))
+        } yield ExitCode.Success
+
+        exitCode.flatMap(ec => {
+          IO(ec)
+        }).recoverWith({
+          case rte: RuntimeException =>
+            //Continue to throw exception until taskToken required
+            if(noTaskToken) throw rte
+            for {
+              _ <- stepFunction.publishFailure(taskToken, exportFailedErrorMessage + s": ${rte.getMessage}")
+            } yield ExitCode.Error
+          case _ =>
+            //Continue to throw exception until taskToken required
+            if(noTaskToken) throw new RuntimeException(exportFailedErrorMessage)
+            for {
+              _ <- stepFunction.publishFailure(taskToken, exportFailedErrorMessage)
+            } yield ExitCode.Error})
```

Comment:
What are the implications for returning `ExitCode.Error` rather than returning an exception wrapped in the IO? I don't know Cats very well, so I don't know if returning the failed IO results in more information in the logs than just returning the error exit code?

Reply:
The `com.monovore.decline.effect.CommandIOApp` which the Main object extends, itself extends the `cats.effect.IOAp`. 

According to `IOApp` documentation (https://typelevel.org/cats-effect/api/cats/effect/IOApp.html): "If the IO raises an error, the stack trace is printed to standard error and sys.exit(1) is called."

So I think it is OK, simply to throw the exception without the need to returning the `ExitCode.Error`